### DISCUSSION TITLE
Help with dynamic parameters

### DISCUSSION THREAD

**Discussion by Alexpathy** [jspsych/jsPsych#2527] [Q&A]
Hi!

I have two questions that I need help with. 

First, I can't define the LastTrialData-property correctly in the var "favorite_city". It renders me a string that just says "undefined" when I run the code.  

```
var test = {
  type: jsPsychHtmlKeyboardResponse,
  stimulus:`
      <div style='width: 700px;'>
      <div style='float: left;'><img src='stimuli_img/DESCRIBE.jpg'></img>
      <p class='small'><strong>Press the A key for option "DESCRIBE"</strong></p></div>
      <div style='float: right;'><img src='stimuli_img/FEEL.jpg'</img>
      <p class='small'><strong>Press the K key for option "FEEL"</strong></p></div>
      `,
  choices: ['a','k'],
  on_finish: function(data){
    // Score the response as correct or incorrect.
    if(jsPsych.pluginAPI.compareKeys(data.response, "k")){
      data.correct = true;
    } else {
      data.correct = false; 
    }
  }
}

var testfeedback = {
  type: jsPsychHtmlKeyboardResponse,
  stimulus: function(){
    // The feedback stimulus is a dynamic parameter because we can't know in advance whether
    // the stimulus should be 'correct' or 'incorrect'.
    // Instead, this function will check the accuracy of the last response and use that information to set
    // the stimulus value on each trial.
    var last_trial_correct = jsPsych.data.get().last(1).values()[0].correct;
    if(last_trial_correct){
      return "FEEL"; // the parameter value has to be returned from the function
    } else {
      return "DESCRIBE"; // the parameter value has to be returned from the function
    }
  }
}

var trialtest = {
  type: jsPsychSurveyText,
  premable: `<img src="stimuli/CFD-BF-037-013-A.jpg" style="width:400px;"></img>`,
  questions: [
    { 
      prompt: function() {  
        // this question prompt is dynamic - the text that is shown 
        // will change based on the participant's earlier response
        var favorite_city = jsPsych.data.getLastTrialData().values()[0].response.last_trial_correct;
        var text = ""+favorite_city+".?"
        return text;
      }, 
      required: true,
      rows: 5,
      columns: 30
    },
  ]
}
timeline.push(test, testfeedback,trialtest);
```

Second question: 
the image is not displaying from the premable: `<img src="stimuli/CFD-BF-037-013-A.jpg" style="width:400px;"></img>`, when I run the code. According to one of the plugin examples I should be able to display images together with questions and prompts. Any thoughts on why the image won't be displayed?

**Comment by jodeleeuw** [comment]
> First, I can't define the LastTrialData-property correctly in the var "favorite_city". It renders me a string that just says "undefined" when I run the code.

It looks like the problem is here:

```js
var favorite_city = jsPsych.data.getLastTrialData().values()[0].response.last_trial_correct;
```

This gets the data from the previous trial (`testfeedback`). The `response` looks like it could be any key, but `response.last_trial_correct` will be `undefined` because there's no property `last_trial_correct` on the `response` in the data. I'm not sure what you are trying to access here, so I'm not sure what to suggest as a fix.

> Second question: the image is not displaying from the premable: <img src="stimuli/CFD-BF-037-013-A.jpg" style="width:400px;"></img>, when I run the code. According to one of the plugin examples I should be able to display images together with questions and prompts. Any thoughts on why the image won't be displayed?

I think this is because the parameter is `preamble` and you have a typo (`premable`).

**Reply by Alexpathy** [reply]
Thank you for your swift reply!

Basically what I want to do is present a participant with two images and let them choose one of the two. 
In the next trial I want to present a text that is conditioned on their previous choice together with a new image and the opportunity to type an answer before continuing (seems SurveyText is the plugin suited for this). 

I tried to create a dynamic prompt that delivers the word "FEEL" or the word "DESCRIBE" based on the choice made during presentation of the variable "test", as participants choose either "a" or "k" during that trial. But that's where my problem is, as my getLastTrialData-function don't find anything. 
Hope my answer is interpretable :) and sorry for my inefficent code as I am just beginning to learn how to program. There is most definently a way to implement the basic conditional logic that I am after in a much smoother way.  

The second problem was due to a typo so it's solved now. Thanks!
